src/holon/models/scenario.py

```diff
@@ -73,105 +76,118 @@ def contracts(self) -> "list[Contract]":
+
         with transaction.atomic():
-            new_scenario = duplicate_model(self, {"cloned_from": scenario_old})
+            new_scenario = duplicate_model_nomutate(self)
+            new_scenario.cloned_from = self
+            new_scenario.save()
 
-            actors = scenario_old.actor_set.all()
             actor_id_to_new_model_mapping = {}
 
-            for actor in actors:
-                actor_id = actor.id
-                actor.original_id = actor_id
-                new_actor = duplicate_model(actor, {"payload": new_scenario})
+            for actor in self.actor_set.all():
+                new_actor = duplicate_model_nomutate(actor)
+                new_actor.payload = new_scenario
+                new_actor.original_id = actor.id
 
-                actor_id_to_new_model_mapping[actor_id] = new_actor
+                actor_id_to_new_model_mapping[actor.id] = new_actor
 
-            contracts = Contract.objects.filter(actor__payload_id=scenario_old.id)
-            for contr in contracts:
-                contr.original_id = contr.id
-                duplicate_model(
-                    contr,
-                    {
-                        "actor": actor_id_to_new_model_mapping[contr.actor.id],
-                        "contractScope": actor_id_to_new_model_mapping[contr.contractScope.id],
-                    },
-                )
+            Actor.objects.bulk_create(actor_id_to_new_model_mapping.values())
 
-            gridnodes = scenario_old.gridnode_set.all()
+            for contract in Contract.objects.filter(actor__payload_id=self.id):
+                new_contract = duplicate_model_nomutate(contract)
+                new_contract.original_id = contract.id
+                new_contract.actor = actor_id_to_new_model_mapping[contract.actor_id]
+                new_contract.contractScope = actor_id_to_new_model_mapping[
+                    contract.contractScope_id
+                ]
+                # Can't bulk create because it's not implemented in django-polymorphic
+                new_contract.save()
 
             gridnode_id_to_new_model_mapping = {}
-            for gridnode in gridnodes:
-                gridnode_id = gridnode.id
-                gridnode.original_id = gridnode_id
-                new_gridnode = duplicate_model(
-                    gridnode,
-                    {
-                        "payload": new_scenario,
-                        "owner_actor": actor_id_to_new_model_mapping[gridnode.owner_actor_id],
-                    },
-                )
-
-                gridnode_id_to_new_model_mapping[gridnode_id] = new_gridnode
-
-                assets = EnergyAsset.objects.filter(gridnode_id=gridnode_id)
-                for asset in assets:
-                    asset.original_id = asset.id
-                    duplicate_model(asset, {"gridnode": new_gridnode})
+            for gridnode in self.gridnode_set.all():
+                new_gridnode = duplicate_model_nomutate(gridnode)
```

Comment:
The only difference here with the duplicate_model function is that attributes are set manually instead of through a dictionary. Does doing it this way have a performance advantage over the previous approach? If not, for readability I would stick with the original `duplicate_model` function, or at least move this code to a function.

Reply:
> The only difference here with the duplicate_model function is that attributes are set manually instead of through a dictionary. Does doing it this way have a performance advantage over the previous approach? If not, for readability I would stick with the original duplicate_model function,

I think the most clear way syntactically to set an attribute is through an assignment to that attribute.

It is also more likely that an IDE or static analysis tool "understands" this code. 

> at least move this code to a function.

I can split it up in functions `copy_actors`, `copy_gridnodes` etc. Probably best to do this in a separate module `clone_scenario.py` rather than scattered over all model classes.

Reply:
Pushed a change to split this code over multiple functions

Reply:
I am also definitely a big proponent of strongly typed over weakly typed programming. However in this case the flexibility and compactness of the current function is useful. I would say change it if it improves performance, otherwise there is no need to change what already works.

Reply:
I have benchmarked the difference.

The scenario_clone() with the intermediate dicationary is 2.24s and direct assignment is 2.17s. This is an average of 10 runs.